### ISSUE TITLE
Churn calculation is slow and I can't know why

### ISSUE THREAD

**Issue by federicoaldunate** [whitesmith/rubycritic#490]
## Context
I'm trying to run rubycritic in a legacy project. Git history might be big
## Issue
It's stuck in `running churn`.
There is no `verbose` option so I cannot understand why is stuck.
```
..........................................................................................................................................................................................................................................................................................................................................................................................................................................................
running churn

```
It would be nice to know how to debug this issue.

**Comment by hemalvarambhia** [comment]
Are there ways of making the computation of churn faster?

**Comment by hemalvarambhia** [comment]
I found this gem by Dan Mayer (see [Churn's Github page](https://github.com/danmayer/churn/blob/main/README.md)) and found it was a lot faster. I use `rubycritic` on some repositories and some take 15+ minutes. Can using the said gem happen?

**Comment by etagwerker** [comment]
@hemalvarambhia Yes, I think that would be a good idea. One question about churn, does it support all the SCMs that RubyCritic supports? (Perforce, Git, Mercurial) 

I'd like to make it faster while not removing support for any SCMs: https://github.com/whitesmith/rubycritic/tree/7de7682cf979116e681b03c02b79eede5e83c855/lib/rubycritic/source_control_systems

Do you think that would be possible? 

I see that `churn` may not have support for Perforce or Mercurial: https://github.com/danmayer/churn/tree/main/lib/churn/scm

I'm still open to a solution that uses `churn` for Git repositories and our current implementation for other SCMs.

Let me know!

**Comment by hemalvarambhia** [comment]
> I'm still open to a solution that uses churn for Git repositories and our current implementation for other SCMs.

I'd love that. I could imagine that we need to refactor the rubycritic to enable integration. I would love to refactor the code to integrate `churn`. To address your concern about not removing support for any SCMs, I imagine, for git, `rubycritic` uses `churn` but falls back to the current implementation for the other SCMs like so:

```ruby
if git
# do the new thing
else
# do the old thing
```

Then, as `churn` adds other SCMs (maybe we contribute to their gem!), we just update the `if` until `else` is no longer executed.

How does that sound?

**Comment by hemalvarambhia** [comment]
@etagwerker 

> I see that churn may not have support for Perforce or Mercurial: https://github.com/danmayer/churn/tree/main/lib/churn/scm

I thought [hg_analyzer](https://github.com/danmayer/churn/blob/main/lib/churn/scm/hg_analyzer.rb) may be the mercurial one. If I were to do this, I'd happily make sure that `churn` can be integrated for our mercurial users. Is that OK?

**Comment by hemalvarambhia** [comment]
Cool. I will try to drive this one. I will have a look at a code of conduct to understand my bounded context :)

**Comment by rishijain** [comment]
A fix for fixing the churn slowness was merged yesterday #535 . Perhaps you can try and see if that speed things up for you @hemalvarambhia .
